### ISSUE TITLE
(jeromq 0.3.5/master) REQ socket receives incomplete (large) message

### ISSUE THREAD

**Issue by andreasweber** [zeromq/jeromq#334]
It seems that you can receive incomplete messages if a disconnect happens during message decoding.
I debugged the jeromq code and found the following behaviour in `V1Decoder/DecoderBase.`

Let's assume we already decoded the message's flags and size, so we are in state `MESSGAGE_READY.`
The message data is now read via `DecoderBase.getBuffer()` in several portions after getting in-events.

Now when there is a disconnect while reading the data, `DecoderBase.stalled()` is called. 

``` java
private boolean stalled()
     //  Check whether there was decoding error.
     if (!next()) {
         return false;
     }
     ...
```

Because `V1Decoder` is in state `MESSGAGE_READY,` the `next()` call leads to a call of `messageReady().`
And here the `inProgress` message with incomplete data is pushed to the sink.

Afterwards, `StreamEngine.error()` is called, which tries to remove incomplete messages via `SessionBase.detach()` resp. `cleanPipes().`
Well, this checks for incomplete messages, but incompletion here means only, that another Message frame is missing (MORE flag). It doesn't check for incompletion in the sense of missing data within a frame. 

I observed this behaviour with `ZMsg.recvMsg(<REQ socket>)`, but IMHO it's a fundamental problem.

Any opinions?

Best regards, 
  Andreas

P.S. It's hard to provide a real unit test for this issue that always fails/works, but I added a simple program that shows the behaviour if you choose msg size big enough.

[TestJeroMqDisconnect.zip](https://github.com/zeromq/jeromq/files/161522/TestJeroMqDisconnect.zip)

**Comment by t-wiser** [comment]
I have been using jeromq 0.3.4 with a DEALER socket and just noticed this issue while checking to see if there was a fix. I had not debugged into the jeromq code, but the behavior is the same.

In my case I was sending a sequence of 1MB packets from a ROUTER (libzmq-based) to a DEALER (jeromq-based) while simulating an unreliable wireless connection. Using wireshark I was able to see that, for example, less than 200KB of the 1MB was received by the DEALER before the connection was lost. After the connection returned, my client code reported the receipt of the full 1MB packet even though only less than 200KB had been received. This resulted in large gaps in the received data.

Putting aside the fact that sending such large packets over an unreliable network is probably not the best idea (long story), I agree with @andreasweber that this is a fundamental problem and would love to see a fix. Until then, I'll have to implement some mitigation strategies.

**Comment by fredoboulo** [comment]
The test with corresponding code was added in PR https://github.com/zeromq/jeromq/pull/435

**Comment by daveyarwood** [comment]
@andreasweber / @t-wiser Are you still experiencing this issue using the latest version of JeroMQ?

**Comment by andreasweber** [comment]
@daveyarwood Test is running successfully now - so from my point of view this seems to be fixed.

**Comment by daveyarwood** [comment]
Great! :+1: 

If there are no objections, I'm going to close this issue. We can reopen it if someone confirms it's still an issue with the current version of JeroMQ.

**Comment by andreasweber** [comment]
That's fine with me - btw, I used latest release 0.4.2 for testing.
